File: formal/coqimp/task05/extraction_hs/Monad.hs
{-# OPTIONS_GHC -cpp -XMagicHash #-}
{- For Hugs, use the option -F"cpp -P -traditional" -}

module Monad where

import qualified Prelude

#ifdef __GLASGOW_HASKELL__
import qualified GHC.Base
#else
-- HUGS
import qualified IOExts
#endif

#ifdef __GLASGOW_HASKELL__
unsafeCoerce :: a -> b
unsafeCoerce = GHC.Base.unsafeCoerce#
#else
-- HUGS
unsafeCoerce :: a -> b
unsafeCoerce = IOExts.unsafeCoerce
#endif

#ifdef __GLASGOW_HASKELL__
type Any = GHC.Base.Any
#else
-- HUGS
type Any = ()
#endif

__ :: any
__ = Prelude.error "Logical or arity value used"

data MonadOps m =
   Build_MonadOps (() -> () -> m -> (Any -> m) -> m) (() -> Any -> m)

bind :: (MonadOps a1) -> a1 -> (a2 -> a1) -> a1
bind monadOps x x0 =
  case monadOps of {
   Build_MonadOps bind0 _ -> unsafeCoerce bind0 __ __ x x0}

ret :: (MonadOps a1) -> a2 -> a1
ret monadOps x =
  case monadOps of {
   Build_MonadOps _ ret0 -> unsafeCoerce ret0 __ x}

data MonadTransformerOps t =
   Build_MonadTransformerOps (() -> (MonadOps Any) -> MonadOps t) (() -> (MonadOps Any) -> () -> Any
                                                                  -> t)

coq_Tops :: (MonadTransformerOps a1) -> (MonadOps a2) -> MonadOps a1
coq_Tops monadTransformerOps h =
  case monadTransformerOps of {
   Build_MonadTransformerOps tops _ -> unsafeCoerce tops __ h}

idM :: MonadOps Any
idM =
  Build_MonadOps (\_ _ x f -> f x) (\_ x -> x)

monad_of_transformer :: (MonadTransformerOps a1) -> MonadOps a1
monad_of_transformer h =
  coq_Tops h idM

optionT :: MonadTransformerOps Any
optionT =
  Build_MonadTransformerOps (\_ h -> Build_MonadOps (\_ _ x f ->
    bind h x (\a -> case a of {
                     Prelude.Just a' -> f a';
                     Prelude.Nothing -> ret h Prelude.Nothing})) (\_ x -> ret h (Prelude.Just x)))
    (\_ h _ x -> bind h x (\a -> ret h (Prelude.Just a)))

optionM :: MonadOps (Prelude.Maybe Any)
optionM =
  monad_of_transformer (unsafeCoerce optionT)

File: formal/coqimp/task01/extraction_ml/Monad.ml

type __ = Obj.t
let __ = let rec f _ = Obj.repr f in Obj.repr f

type 'm coq_MonadOps = { bind : (__ -> __ -> 'm -> (__ -> 'm) -> 'm); ret : (__ -> __ -> 'm) }

(** val bind : 'a1 coq_MonadOps -> 'a1 -> ('a2 -> 'a1) -> 'a1 **)

let bind monadOps x x0 =
  let { bind = bind0; ret = _ } = monadOps in Obj.magic bind0 __ __ x x0

(** val ret : 'a1 coq_MonadOps -> 'a2 -> 'a1 **)

let ret monadOps x =
  let { bind = _; ret = ret0 } = monadOps in Obj.magic ret0 __ x

type 't coq_MonadTransformerOps = { coq_Tops : (__ -> __ coq_MonadOps -> 't coq_MonadOps);
                                    lift : (__ -> __ coq_MonadOps -> __ -> __ -> 't) }

(** val coq_Tops : 'a1 coq_MonadTransformerOps -> 'a2 coq_MonadOps -> 'a1 coq_MonadOps **)

let coq_Tops monadTransformerOps h =
  let { coq_Tops = tops; lift = _ } = monadTransformerOps in Obj.magic tops __ h

(** val idM : __ coq_MonadOps **)

let idM =
  { bind = (fun _ _ x f -> f x); ret = (fun _ x -> x) }

(** val monad_of_transformer : 'a1 coq_MonadTransformerOps -> 'a1 coq_MonadOps **)

let monad_of_transformer h =
  coq_Tops h idM

(** val optionT : __ coq_MonadTransformerOps **)

let optionT =
  { coq_Tops = (fun _ h -> { bind = (fun _ _ x f ->
    bind h x (fun a -> match a with
                       | Some a' -> f a'
                       | None -> ret h None)); ret = (fun _ x -> ret h (Some x)) }); lift = (fun _ h _ x ->
    bind h x (fun a -> ret h (Some a))) }

(** val optionM : __ option coq_MonadOps **)

let optionM =
  monad_of_transformer (Obj.magic optionT)
